### ISSUE TITLE
Better RobotMap Design?

### ISSUE THREAD

**Issue by ryanlarkin** [FRC2706/2016-2706-Rubiks#3]
We have different robots: practice, competition, and possibly simulation.  Some of the calibration and channel values may be different from robot to robot.  Currently `RobotMap`, which has most of this information, uses public constants, which other code uses for mechanisms and subsystems.  It is possible to determine whether the robot is the practice robot or the competition robot using a `DigitalInput` placed on both robots.  The problem with this is that WPI must initialize all of it's code before we can determine the value of the `DigitalInput`.  Since `RobotMap` uses constants, we cannot change  them when we know when we know which robot we are using.  My question is, without using reflection and preferably without having to write an if statement whenever we need a `RobotMap` value, or having to add a getter method everytime we need to add a new `RobotMap` value, how can we have multiple `RobotMap` values for different robots?

tl:dr: How do we make the RobotMap support multiple robots with different configurations, without being hacky or overly redundant?

**Comment by ryanlarkin** [comment]
So it looks like we will be using @georgetzavelas's constant class from last year.  To change the code, if it does not change depending on the robot, then you an just add it to the constants class like this:

```
public static final int NAME_OF_PROPERTY = "NAME_OF_PROPERTY_FOR_CONFIG_FILE";
...
public Constants() {
    defaults.put(NAME_OF_PROPERTY, "Defualt value");
}
```

If it is robot dependent, you must also put it on the robot properties file.

**Comment by ryanlarkin** [comment]
It also seems like it might require less code to use JSON.

**Comment by ryanlarkin** [comment]
I think the simplest way is actually to set the constants be reading a file that identifies the robot type and setting the constants from there.  Here's how:

Accessing constants with RobotMap is the same as usual, however, to add a new constant, you must first add the constant to RobotMap with new value.  Then create the same constant into each nested class, with the value for each robot.  Next in the static block, add a local variable of the same type, initialized to 0.  Go to the switch statement and set the local variable to the nested class version of the constant.  Finally, go to the last part of the static block and set the RobotMap constant to the local variable.

**Comment by eandr127** [comment]
* Accessing constants with RobotMap is the same as usual,
	 * however, to add a new constant, you must first add the constant
	 * to RobotMap with new value.  Then create the same constant
	 * into each nested class, with the value for each robot.
	 * Next in the static block, add a local variable of the same type, 
	 * initialized to 0.  Go to the switch statement and set the local variable
	 * to the nested class version of the constant.  Finally, go to the last part
	 * of the static block and set the RobotMap constant to the local variable.

**Comment by kevlam2706** [comment]
Nice work @eandr127, sorry I never actually said that to you before.

**Comment by ryanlarkin** [comment]
Thank you.
